### ISSUE TITLE
.Net Core 3.1 - Incorrect matching of endpoint template

### ISSUE THREAD

**Issue by gurutcr** [dotnet/aspnetcore#40070]
It looks like the endpoint pattern is matching greedy instead of going down to the next best match. I can also see that it totally ignores the attribute routing specified on the controller action. Please see the code snippets below.

**startup.cs**
```
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllerRoute(
                    name: "one",
                    pattern: "{CompanyType}/{CompanyId}/{controller}/{EmployeeUuid:Guid}/{action}");
                endpoints.MapControllerRoute(
                    name: "two",
                    pattern: "{CompanyType}/{CompanyId}/{controller}/{action}");
            });
```

**TestController.cs**
```
namespace RoutingTest.Controllers
{
    [Route("{CompanyType}/{CompanyId}/[controller]")]
    public class TestController : Controller
    {
        public TestController() { }

        [HttpGet("{EmployeeId:Guid}/[action]", Name = "Employee_Open")]
        public IActionResult OpenEmployee(bool edit = false, CancellationToken ct = default)
        {
            return View("Employee");
        }
        [HttpPost("[action]", Name = "Employee_Save")]
        public IActionResult SaveEmployee(Employee model, CancellationToken ct = default)
        {
            return RedirectToRoute("Employee_Open", new { CompanyType = "hr", CompanyId = "776cfe52-d39a-411d-88f3-c97e6895eb63", EmployeeId = Guid.NewGuid() });
        }
    }
}
```

**Currently  loaded page's URL.**

`http://localhost:5000/hr/776cfe52-d39a-411d-88f3-c97e6895eb63/Test/1d3843a4-7409-4f19-affc-de9aa47f48db/OpenEmployee?edit=true`

**Employee.cshtml**

```
@model Employee

<form asp-controller="test" asp-action="saveemployee">
    @Html.AntiForgeryToken()

    <div>
        <div>
            <button asp-controller="test" asp-action="saveemployee" type="submit">
                <i class="fas fa-edit"></i>Save
            </button>
            <a asp-route="Employee_Open" asp-route-employeeid="@Model.uuid" asp-route-edit="true">
                <i class="fas fa-edit"></i> Edit
            </a>
        </div>
    </div>
</form>

```

**Generate HTML of the save button**

`<button type="submit" formaction="/hr/776cfe52-d39a-411d-88f3-c97e6895eb63/test/1d3843a4-7409-4f19-affc-de9aa47f48db/saveemployee">Save</button>`

As you can see the generated **formaction** url inserted the employee id before the action name though it is not there as a token in the specified attribute routing. This happens because it matches the first endpoint defined in the startup class instead of going down to the second endpoint. Is this a bug in .net core 3.1? Or is there something wrong in my code or approach?

**Comment by javiercn** [comment]
@greenEkatherine thanks for contacting us.

I think there's a gap in your understanding of how link generation works. What's happening is that when you are generating a link in the context of the rendered page, there are a set of ambient values that are taken into account:
```
CompanyType=hr
CompanyId=776cfe52-d39a-411d-88f3-c97e6895eb63
controller=test
EmployeeUuid=1d3843a4-7409-4f19-affc-de9aa47f48db
action=OpenEmployee
```

Since you are providing `asp-controller="test" asp-action="saveemployee"`, routing is able to generate a template using the link below.
```
"{CompanyType}/{CompanyId}/{controller}/{EmployeeUuid:Guid}/{action}");
```

There is no defined order/priority in link generation. The moment we find a template that can generate a URL, we do so. If you want to use a specific route template, then you need to give your route a name and use that name during link generation.

**Comment by javiercn** [comment]
> Is attribute routing completely ignored or overridden here by endpoint template? I couldn't find any documentation that explains how the endpoints and attribute routing works in combination.

It's neither overridden nor ignored. Routes defined via attributes are not special in any way, they are just endpoints, the same as the ones defined when you use a conventional route. When multiple routes can generate a link for an action, the "first" one that  happens to be checked first wins, in this case, the conventional route happens to be probed first, and hence why you get that link.

If you want to make sure that a specific template is used, then you need to give it a name. Otherwise for what matters to routing, both options are equally acceptable, so the one that happens to be tried first is the one that gets picked.

**Comment by gurutcr** [comment]
> If you want to make sure that a specific template is used, then you need to give it a name. 

I didn't quite understand this. I know we can give name to endpoints and for attribute routes too. Are you saying to call the route by name in button or anchor tags to hit that route specifically?

From your explanation it looks like the conventional routes are tried first, so isn't it kind of overriding the attribute routing?

On another note, I had to define endpoints apart from the default one (that comes with the .net project template) just to make sure the route values are used while generating the links. I mean, if there is no endpoint defined in the startup class that has a route token (EmployeeId in our example) in it, then it just don't get replaced in the attribute route that has this token. I don't know if I explained it well.

So in other words, to solve the issue in my example, what if I just don't define any endpoints and just use attribute routing? Will the ambient values be used in link generation then?

**Comment by javiercn** [comment]
> I didn't quite understand this. I know we can give name to endpoints and for attribute routes too. Are you saying to call the route by name in button or anchor tags to hit that route specifically?

Yes, you should be using a route name if you want a specific route template is used.

> From your explanation it looks like the conventional routes are tried first, so isn't it kind of overriding the attribute routing?

No, there's no defined order and the chosen template we use to generate the link and the one that gets used might change in the future. That's why you need to use the name to make sure your desired route template is used.

> On another note, I had to define endpoints apart from the default one (that comes with the .net project template) just to make sure the route values are used while generating the links. I mean, if there is no endpoint defined in the startup class that has a route token (EmployeeId in our example) in it, then it just don't get replaced in the attribute route that has this token. I don't know if I explained it well.

> So in other words, to solve the issue in my example, what if I just don't define any endpoints and just use attribute routing? Will the ambient values be used in link generation then?

I'm not sure I get the first paragraph, but I can answer your last question which is, ambient values will be used, there's "no difference" between endpoints defined from attributes or conventions (there are some differences, but not to the matching/link generation process)

**Comment by gurutcr** [comment]
@javiercn I totally appreciate your time. I felt like my questions were answered by the best person to answer. I will try both suggestions you made, (1) targeting actions with route name and (2) just having attribute routing and still have the ambient values used in link generation. 

If at all I find any issues, can I comment on the same thread?

**Comment by javiercn** [comment]
@gurutcr thanks for the kind comments. 

> If at all I find any issues, can I comment on the same thread?

It would be best if you file a separate issue and add the necessary context there, otherwise it's easy for us to get confused.
